### PR TITLE
[WIP] Reworked ability to edit invoices, as well as many fixes and smaller changes.

### PR DESCRIPTION
**Pull request by Anuril**
This draft PR is part of a larger overhaul of the invoicing of Fossbilling. 

Here is a list of current changes:

- Changed the Invoice Detail view in the Admin area to be more condensed
grafik

- Removed the ability to edit most of the invoice details.
- Modifying invoice date, due date and paid date is only possible as long as the invoice is unpaid. 
grafik

- Added the ability to refresh the invoice's client data from the client record
grafik

- Invoices can't be deleted anymore if the client had the possibility to see it (Meaning: The invoice has the state of "approved")
- Instead, invoices can be revoked if they're unpaid. Paid, refunded, cancelled or revoked Invoices can't be deleted. 
- If an unpaid invoice isn't approved, it can be deleted just as before.
- Revoking an Invoice removes it from the associated order, making the order "available" again for manual or automated renewal.
grafik

- Added an Archive tab to the invoice list to quickly see revoked Invoices. (And so they don't show up in the invoice list)
grafik

- Fixed the Missing Price field in the Invoice item list, changed formatting so it's more consistent
grafik

- Changed colors (and batch formats) for invoice states so they are consistent between admin and client interfaces
grafik

- As most helpfully implemented by @gilesytheking in #1992, I've changed the due date / paid date fields in the tables, as well as changing many smaller inconsistencies around ```format_date```
grafik
grafik

- Added "Refunded" and "Revoked" tabs to the client Invoice overview (Only displayed if the client has refunded or revoked invoices)
grafik

- Changed the wording for most Client-Side Things around refunded Invoices to "Credit Note" 
grafik

Todos

- [ ] Filtering for invoices is broken unless and until a Client is selected. 
grafik
grafik

- [ ] VAT % and Amounts are not stored in the invoice. invoice item price is the same as total in the database. Currently, it's a workaround that calculates the values in Situ, which has to be fixed.
- [ ] For the Revoke functionality, no tests exist yet, and while I'm at it, I'll go over the other existing tests for the Invoice Module

### CODE REVIEW

File: src/modules/Invoice/Api/Admin.php [hunk 1]
```diff
@@ -217,8 +232,17 @@ public function item_delete($data)
     public function delete($data)
     {
         $model = $this->_getInvoice($data);
-
-        return $this->getService()->deleteInvoiceByAdmin($model);
+        // check if invoice is unpaid and not approved
+        if ($model->status == 'unpaid') {
+            // if invoice is approved, then revoke it
+            if ($model->approved == '1') {
```
Comment (BelleNottelling) [inline src/modules/Invoice/Api/Admin.php:238]: Same logic regardless of the conditions in the if statement?

**Comment by Anuril** [comment]
This is still WIP and will need quite some work but with this PR, this changes:

- [ ] Proforma invoice functionality
- [ ] Re-Integrated Quantities in the order form
- [ ] Made Addons fully functional Products
- [ ] Made Orders cancellable by the user
- [ ] Made invoice numbers saved to the database instead of being dynamically printed with the settings information (Legal Requirement in most states, certainly EU area and Switzerland.)
- [ ] Made the refund logic switch actually do stuff.
- [ ] Refunding now adds funds back to customer account. (May need to be removed, not yet coupled with payment gateways) 
- [ ] Made the order form more structured 
- [ ] Started work on rewriting the invoice approval functionality which does not really do a lot right now.

New Order form:  
grafik

New Tab "Credit Notes"
grafik

**Comment by Anuril** [comment]
There will need to be a refactoring of this code as well, and many legacy tests deleted and rewritten, so this isn't going to be a quick merge.

**Comment by BelleNottelling** [comment]
Looks like you've got quite a bit of feature creep going on in this PR.

I would suggest pulling the fixes & front-end improvements out of this so we can get them merged sooner rather than later while also making this one more straightforward to review later down the line once you've gotten it completed.

> and many legacy tests deleted and rewritten

That's not 100% true. A fair few have been deleted, however none of them have been rewritten IIRC. Just automated refactorings using Rector. Chances are if you ran rector and PHP-CS-Fixer against this branch those tests should then equal how they are correctly written in our `main` branch.

**Comment by Anuril** [comment]
I meant the tests yet need to be written 🤣

**Comment by Anuril** [comment]
> Looks like you've got quite a bit of feature creep going on in this PR.

Yes and no. There is feature creep but it's not easy to disentangle because there's a close link between orders an invoices. 

I can certainly see some fixes (like the invoice can't be filtered fix) which is just removing the typeful check being implemented quicker, but this PR stems from me starting to fix one thing and noticing that it depends on the next, and that depends on the next again, so I can't see a way around having this PR patching the whole Order & Invoicing Part. At this stage, it's still a "I'm fixing what I see and either completing functionality that was present in the UI but didn't work or restoring functionality that has been working in the past but doesn't work anymore.

The most important aspects of that were: 
- Refund logic which is broken / doesn't exist
- Quantity is supported in the backend but not in the frontend
- "Proforma" functionality which was not clearly positioned -> Changed it so invoice can have three document types: proforma, invoice and creditnote.

What I've yet to look at is the approval logic - This setting in the invoice settings only works on one specific invoice type: 
grafik
The "add funds"-invoice the client can generate. (See [here](https://github.com/FOSSBilling/FOSSBilling/blob/e50abd1ddc3648fd81b88ce55e8512b89c01c6bb/src/modules/Invoice/Service.php#L1100).) But that is only until the next function is called by the calling function [in Invoice/Api/Client.php->funds_invoice](https://github.com/FOSSBilling/FOSSBilling/blob/e6555658afc8d0f3c2c5700715830d242525f7ce/src/modules/Invoice/Api/Client.php#L143), where directly after calling generateFundsInvoice(), approveInvoice is called, making the check moot. 

But nevertheless, the approveInvoice functionality is crucial in every step of the invoice creation process, as well as in the renewal process, and gets called from 8 different functions. 

It also doesn't do what one would expect, as it just sets the flag to approved and tries to pay the invoice with credits. (tryPayWithCredits) 

Because I f'd up the renewal once before and caused #1969, I've (roughly) documented the whole renewal procedure: 
Renewal_logic drawio

and I plan to continue that diagram, adding most creation functions too. 

The next step is then to reduce complexity wherever possible, [like here](https://github.com/FOSSBilling/FOSSBilling/blob/e6555658afc8d0f3c2c5700715830d242525f7ce/src/modules/Invoice/ServiceInvoiceItem.php#L296) where [setUnpaidInvoice](https://github.com/FOSSBilling/FOSSBilling/blob/e6555658afc8d0f3c2c5700715830d242525f7ce/src/modules/Order/Service.php#L1337) literally gets called to save $order->unpaid_invoice_id = $proforma->id and $order->updated_at, and then the order is stored. And then it's stored [again here.](https://github.com/FOSSBilling/FOSSBilling/blob/e6555658afc8d0f3c2c5700715830d242525f7ce/src/modules/Invoice/ServiceInvoiceItem.php#L319)
Which I'm almost entirely sure is unnecessary. 

So TL;DR: I'm open to suggestions what of the functionality we could split out and already implement now, but currently, this is a PR where I put my WIP of refactoring the Invoice & Order overhaul so there is visibility and the possibility of discussion.
